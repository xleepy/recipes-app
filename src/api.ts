// types from https://developer.edamam.com/edamam-docs-recipe-api

type ImageInfo = {
  url?: string;
  width?: string;
  height?: string;
};

type ImageModel = "THUMBNAIL" | "SMALL" | "REGULAR" | "LARGE";

type InlineModel = Record<ImageModel, ImageInfo>;

type Ingredient = {
  text?: string;
  quantity?: number;
  measure?: string;
  food?: string;
  weight?: number;
  foodId?: string;
};

type NutrientsInfo = Record<string, any>;

type Digest = {
  label?: string;
  tag?: string;
  schemaOrgTag?: string;
  total?: number;
  hasRDI?: boolean;
  daily?: number;
  unit?: string;
  sub?: Digest;
};

export type Recipe = {
  uri?: string;
  label?: string;
  image?: string;
  images?: InlineModel;
  source?: string;
  url?: string;
  shareAs?: string;
  yield?: number;
  dietLabels?: string[];
  healthLabels?: string[];
  cautions?: string[];
  ingredientLines?: string[];
  ingredients?: Ingredient[];
  calories?: number;
  glycemicIndex?: number;
  totalCO2Emissions?: number;
  co2EmissionsClass?: string;
  totalWeight?: number;
  cuisineType?: string[];
  mealType?: string[];
  dishType?: string[];
  instructions?: string[];
  tags?: string[];
  externalId?: string;
  totalNutrients?: NutrientsInfo;
  totalDaily?: NutrientsInfo;
  digest?: Digest;
};

type Link = {
  href?: string;
  title?: string;
};

type Links = {
  self?: Link;
  next?: Link;
};

export type Hit = {
  recipe: Recipe;
  _links?: Links;
};

export type ApiResponse = {
  from?: number;
  to?: number;
  count?: number;
  _links?: Links;
  hits?: Hit[];
};
