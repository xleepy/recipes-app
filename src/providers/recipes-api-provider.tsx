import { createContext, ComponentChildren } from 'preact';
import { Configuration, RecipesApi } from '../api';
import { useContext, useMemo } from 'preact/hooks';

type RecipesProviderContextValue = RecipesApi | null;

const RecipesProviderContext = createContext<RecipesProviderContextValue>(null);

export const RecipesApiProvider = ({
  children,
}: {
  children: ComponentChildren;
}) => {
  const api = useMemo(() => {
    return new RecipesApi(
      new Configuration({ apiKey: import.meta.env.VITE_APP_ID })
    );
  }, []);
  return (
    <RecipesProviderContext.Provider value={api}>
      {children}
    </RecipesProviderContext.Provider>
  );
};

export const useRecipesApi = () => {
  const context = useContext(RecipesProviderContext);
  if (!context) {
    throw new Error('Api not provided');
  }
  return context;
};
