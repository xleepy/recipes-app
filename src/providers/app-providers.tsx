import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ComponentChildren } from 'preact';
import { RecipesApiProvider } from './recipes-api-provider';
import { HashRouter } from 'react-router-dom';

type AppProvidersProps = {
  children: ComponentChildren;
};

const queryClient = new QueryClient();

export const AppProviders = ({ children }: AppProvidersProps) => {
  return (
    <QueryClientProvider client={queryClient}>
      <RecipesApiProvider>
        <HashRouter>{children}</HashRouter>
      </RecipesApiProvider>
    </QueryClientProvider>
  );
};
