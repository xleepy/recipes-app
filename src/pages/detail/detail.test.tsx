import { render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import DetailPage from '.';
import { MemoryRouter } from 'react-router-dom';
import { RecipesApiProvider } from '../../providers/recipes-api-provider';
import { GetRecipeInformation200Response } from '../../api';

vi.mock('../../api/apis/RecipesApi', () => {
  const MockInstance = vi.fn(() => {
    return {
      getRecipeInformation: vi.fn(() => {
        return Promise.resolve<GetRecipeInformation200Response>({
          title: 'test',
          imageType: 'png',
          image: 'none',
          id: 1,
          servings: 1,
          vegan: true,
          whole30: false,
          weightWatcherSmartPoints: 1,
          veryPopular: false,
          veryHealthy: true,
          vegetarian: true,
          sustainable: true,
          summary: 'lorem ipsum',
          spoonacularScore: 1,
          spoonacularSourceUrl: 'test',
          sourceUrl: 'test',
          sourceName: 'test',
          readyInMinutes: 15,
          pricePerServing: 1,
          occasions: [],
          lowFodmap: false,
          license: 'test',
          ketogenic: false,
          instructions: 'test',
          healthScore: 12,
          glutenFree: false,
          gaps: 'test',
          extendedIngredients: new Set(),
          dishTypes: [],
          diets: [],
          dairyFree: false,
          cuisines: [],
          creditsText: 'test',
          cheap: false,
          analyzedInstructions: [{ steps: [] }],
          aggregateLikes: 1,
          winePairing: {
            pairedWines: [],
            pairingText: 'test',
            productMatches: new Set(),
          },
        });
      }),
    };
  });
  return {
    RecipesApi: MockInstance,
  };
});

describe('Detail tests', () => {
  it('should render detail', async () => {
    const { asFragment, queryByTestId } = render(
      <MemoryRouter>
        <RecipesApiProvider>
          <DetailPage />
        </RecipesApiProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(queryByTestId('detail')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
