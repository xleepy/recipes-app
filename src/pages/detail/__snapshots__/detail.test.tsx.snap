// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Detail tests > should render detail 1`] = `
<DocumentFragment>
  <div
    class="_detail_21518f"
    data-testid="detail"
  >
    <img
      src="none"
    />
    <div>
      test
    </div>
    <div>
      test
    </div>
  </div>
</DocumentFragment>
`;
